solver/llbsolver/ops/diff.go

```diff
@@ -0,0 +1,105 @@
+package ops
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/moby/buildkit/worker"
+	"github.com/pkg/errors"
+
+	"github.com/moby/buildkit/cache"
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/solver"
+	"github.com/moby/buildkit/solver/llbsolver"
+	"github.com/moby/buildkit/solver/pb"
+	digest "github.com/opencontainers/go-digest"
+)
+
+const diffCacheType = "buildkit.diff.v0"
+
+type diffOp struct {
+	op     *pb.DiffOp
+	worker worker.Worker
+}
+
+func NewDiffOp(v solver.Vertex, op *pb.Op_Diff, w worker.Worker) (solver.Op, error) {
+	if err := llbsolver.ValidateOp(&pb.Op{Op: op}); err != nil {
+		return nil, err
+	}
+	return &diffOp{
+		op:     op.Diff,
+		worker: w,
+	}, nil
+}
+
+func (d *diffOp) CacheMap(ctx context.Context, group session.Group, index int) (*solver.CacheMap, bool, error) {
+	dt, err := json.Marshal(struct {
+		Type string
+		Diff *pb.DiffOp
+	}{
+		Type: diffCacheType,
+		Diff: d.op,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+
+	cm := &solver.CacheMap{
+		Digest: digest.Digest(dt),
+		Deps: make([]struct {
+			Selector          digest.Digest
+			ComputeDigestFunc solver.ResultBasedCacheFunc
+			PreprocessFunc    solver.PreprocessFunc
+		}, 2),
+	}
+
+	return cm, true, nil
+}
+
+func (d *diffOp) Exec(ctx context.Context, g session.Group, inputs []solver.Result) ([]solver.Result, error) {
+	var curInput int
+
+	var lowerRef cache.ImmutableRef
+	var lowerRefID string
+	if d.op.Lower.Input != pb.Empty {
```

Comment:
`if lower == -1 then return upper`?

Same could probably be directly in client as well.

Reply:
As mentioned in my other comment, there is currently a desirable difference between `Diff(Scratch(), A))` and `A` (the diff with scratch can't include any deletions present in `A` since there's nothing to delete from scratch), so will leave this as is for now pending the outcome of that discussion.

Reply:
Updated with this now since we are switching to supporting only unmerge behavior instead of both for now.